src/source/Crypto/Dtls_mbedtls.c

```diff
@@ -0,0 +1,739 @@
+
+    CHK_STATUS(createIOBuffer(DEFAULT_MTU_SIZE, &pDtlsSession->pReadBuffer));
+    pDtlsSession->timerQueueHandle = timerQueueHandle;
+    pDtlsSession->timerId = UINT32_MAX;
+    pDtlsSession->sslLock = MUTEX_CREATE(TRUE);
+    pDtlsSession->dtlsSessionCallbacks = *pDtlsSessionCallbacks;
+    if (certificateBits == 0) {
+        certificateBits = GENERATED_CERTIFICATE_BITS;
+    }
+
+    if (certCount == 0) {
+        CHK_STATUS(createCertificateAndKey(
+              certificateBits, 
+              generateRSACertificate, 
+              &pDtlsSession->certificates[0].cert,
+              &pDtlsSession->certificates[0].privateKey
+        ));
+        pDtlsSession->certificateCount = 1;
+    } else {
+        for (i = 0; i < certCount; i++) {
+            CHK_STATUS(copyCertificateAndKey(
+                  (mbedtls_x509_crt*) pRtcCertificates[i].pCertificate,
+                  (mbedtls_pk_context*) pRtcCertificates[i].pPrivateKey,
+                  &pDtlsSession->certificates[i]
+            ));
+            // in case of a failure in between, we'll only free up to current position
+            pDtlsSession->certificateCount++;
+        }
+    }
+
+    // Generate and store the certificate fingerprints
+    for (i = 0; i < pDtlsSession->certificateCount; i++) {
+        pCertInfo = pDtlsSession->certificates + i;
+        CHK_STATUS(dtlsCertificateFingerprint(&pCertInfo->cert, pCertInfo->fingerprint));
+    }
+    *ppDtlsSession = pDtlsSession;
+
+CleanUp:
+
+    CHK_LOG_ERR(retStatus);
+
+    if (STATUS_FAILED(retStatus) && pDtlsSession != NULL) {
+        freeDtlsSession(&pDtlsSession);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS freeDtlsSession(PDtlsSession *ppDtlsSession)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT32 i;
+    PDtlsSessionCertificateInfo pCertInfo;
+    PDtlsSession pDtlsSession;
+
+    CHK(ppDtlsSession != NULL, STATUS_NULL_ARG);
+
+    pDtlsSession = *ppDtlsSession;
+    CHK(pDtlsSession != NULL, retStatus);
+
+    if (pDtlsSession->timerId != UINT32_MAX) {
+        timerQueueCancelTimer(pDtlsSession->timerQueueHandle, pDtlsSession->timerId, (UINT64) pDtlsSession);
+    }
+
+    for (i = 0; i < pDtlsSession->certificateCount; i++) {
+        pCertInfo = pDtlsSession->certificates + i;
+        freeCertificateAndKey(&pCertInfo->cert, &pCertInfo->privateKey);
+    }
+    mbedtls_entropy_free(&pDtlsSession->entropy);
+    mbedtls_ctr_drbg_free(&pDtlsSession->ctrDrbg);
+    mbedtls_ssl_config_free(&pDtlsSession->sslCtxConfig);
+    mbedtls_ssl_free(&pDtlsSession->sslCtx);
+
+    freeIOBuffer(&pDtlsSession->pReadBuffer);
+    if (IS_VALID_MUTEX_VALUE(pDtlsSession->sslLock)) {
+        MUTEX_FREE(pDtlsSession->sslLock);
+    }
+    SAFE_MEMFREE(*ppDtlsSession);
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+INT32 dtlsSessionSendCallback(PVOID customData, const unsigned char *pBuf, ULONG len)
+{
+  STATUS retStatus = STATUS_SUCCESS;
```

Comment:
spacing is off - 2 instead of 4 spaces below this

Reply:
Done